horovod/spark/common/store.py

```diff
@@ -275,66 +262,83 @@ def _get_full_path_or_default(self, path, default_key):
+                yield tmpdir
+        return local_run_path
 
-    def get_filesystem(self):
+    def get_localized_path(self, path):
         raise NotImplementedError()
 
-    @classmethod
-    def matches(cls, path):
-        return path.startswith(cls.filesystem_prefix())
+    def get_full_path(self, path):
+        raise NotImplementedError()
 
-    @classmethod
-    def filesystem_prefix(cls):
+    def get_full_path_fn(self):
         raise NotImplementedError()
 
+    @property
+    def fs(self):
+        raise NotImplementedError()
 
-class LocalStore(FilesystemStore):
-    """Uses the local filesystem as a store of intermediate data and training artifacts."""
 
-    FS_PREFIX = 'file://'
+class FilesystemStore(AbstractFilesystemStore):
+    """Concrete filesystems store that delegates to `fsspec`."""
 
     def __init__(self, prefix_path, *args, **kwargs):
-        self._fs = pa.LocalFileSystem()
-        super(LocalStore, self).__init__(prefix_path, *args, **kwargs)
+        self._fs, self.protocol = FilesystemStore._get_fs_and_protocol(prefix_path)
+        super().__init__(prefix_path, *args, **kwargs)
 
-    def path_prefix(self):
-        return self.FS_PREFIX
+    def sync_fn(self, run_id):
+        run_path = self.get_run_path(run_id)
 
-    def get_filesystem(self):
-        return self._fs
+        def fn(local_run_path):
+            self.fs.put(local_run_path, run_path, recursive=True)
 
-    def get_local_output_dir_fn(self, run_id):
-        run_path = self.get_localized_path(self.get_run_path(run_id))
+        return fn
 
-        @contextlib.contextmanager
-        def local_run_path():
-            if not os.path.exists(run_path):
-                try:
-                    os.makedirs(run_path, mode=0o755)
-                except OSError as e:
-                    # Race condition from workers on the same host: ignore
-                    if e.errno != errno.EEXIST:
-                        raise
-            yield run_path
+    def get_localized_path(self, path):
+        _, lpath = split_protocol(path)
+        return lpath
 
-        return local_run_path
+    def get_full_path(self, path):
+        return self.get_full_path_fn()(path)
 
-    def sync_fn(self, run_id):
-        run_path = self.get_localized_path(self.get_run_path(run_id))
+    def get_full_path_fn(self):
+        def get_path(path):
+            protocol, _ = split_protocol(path)
+            if protocol is not None:
+                return path
+            return pathlib.Path(os.path.abspath(path)).as_uri()
+        return get_path
 
-        def fn(local_run_path):
-            # No-op for LocalStore since the `local_run_path` will be the same as the run path
-            assert run_path == local_run_path
-        return fn
+    @property
+    def fs(self):
+        return self._fs
+
+    @staticmethod
+    def _get_fs_and_protocol(url):
+        protocol, path = split_protocol(url)
+        fs = fsspec.filesystem(protocol)
```

Comment:
Can storage_options be provided as second argument [fsspec.filesystem(protocol, **storage_options)] ? This will help provide authentication options to fsspec.